tools/lsp/common/document_cache.rs

```diff
@@ -0,0 +1,419 @@
+        >,
+    >,
+>;
+
+pub struct CompilerConfiguration {
+    pub include_paths: Vec<std::path::PathBuf>,
+    pub library_paths: HashMap<String, std::path::PathBuf>,
+    pub style: Option<String>,
+    pub open_import_fallback: OpenImportFallback,
+    pub resource_url_mapper:
+        Option<Rc<dyn Fn(&str) -> Pin<Box<dyn Future<Output = Option<String>>>>>>,
+}
+
+impl Default for CompilerConfiguration {
+    fn default() -> Self {
+        let mut cc = default_cc();
+
+        Self {
+            include_paths: std::mem::take(&mut cc.include_paths),
+            library_paths: std::mem::take(&mut cc.library_paths),
+            style: std::mem::take(&mut cc.style),
+            open_import_fallback: None,
+            resource_url_mapper: std::mem::take(&mut cc.resource_url_mapper),
+        }
+    }
+}
+
+impl CompilerConfiguration {
+    fn to_compiler_configuration(
+        mut self,
+    ) -> (i_slint_compiler::CompilerConfiguration, OpenImportFallback) {
+        let mut result = default_cc();
+        result.include_paths = std::mem::take(&mut self.include_paths);
+        result.library_paths = std::mem::take(&mut self.library_paths);
+        result.style = std::mem::take(&mut self.style);
+        result.resource_url_mapper = std::mem::take(&mut self.resource_url_mapper);
+
+        (result, self.open_import_fallback)
+    }
+}
+
+/// A cache of loaded documents
+pub struct DocumentCache {
+    type_loader: TypeLoader,
+    open_import_fallback: OpenImportFallback,
+    source_file_versions: Rc<RefCell<SourceFileVersionMap>>,
+}
+
+pub fn document_cache_parts_setup(
+    compiler_config: &mut i_slint_compiler::CompilerConfiguration,
+    open_import_fallback: OpenImportFallback,
+    initial_file_versions: SourceFileVersionMap,
+) -> (OpenImportFallback, Rc<RefCell<SourceFileVersionMap>>) {
+    let source_file_versions = Rc::new(RefCell::new(initial_file_versions));
+    DocumentCache::wire_up_import_fallback(
+        compiler_config,
+        open_import_fallback,
+        source_file_versions,
+    )
+}
+
+impl DocumentCache {
+    fn wire_up_import_fallback(
+        compiler_config: &mut i_slint_compiler::CompilerConfiguration,
+        open_import_fallback: OpenImportFallback,
+        source_file_versions: Rc<RefCell<SourceFileVersionMap>>,
+    ) -> (OpenImportFallback, Rc<RefCell<SourceFileVersionMap>>) {
+        let sfv = source_file_versions.clone();
+        if let Some(open_import_fallback) = open_import_fallback.clone() {
+            compiler_config.open_import_fallback = Some(Rc::new(move |file_name: String| {
+                let flfb = open_import_fallback(file_name.clone());
+                let sfv = sfv.clone();
+                Box::pin(async move {
+                    flfb.await.map(|r| {
+                        r.map(|(v, c)| {
+                            let path = PathBuf::from(&file_name);
+                            sfv.borrow_mut().insert(path, v);
+                            c
```

Comment:
if `r` is none, we should remove it from the source_file_versions

Reply:
True